File: API/1999_Color_Zerolag_TriX_OSMA/PY/color_zerolag_trix_osma_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import Trix, ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy


class color_zerolag_trix_osma_strategy(Strategy):

    def __init__(self):
        super(color_zerolag_trix_osma_strategy, self).__init__()

        self._trix_period = self.Param("TrixPeriod", 14) \
            .SetDisplay("TRIX Period", "TRIX calculation period", "Indicator")
        self._signal_period = self.Param("SignalPeriod", 9) \
            .SetDisplay("Signal Period", "Signal line EMA period", "Indicator")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles", "General")

        self._prev_osma = 0.0
        self._prev_prev_osma = 0.0
        self._count = 0

    @property
    def TrixPeriod(self):
        return self._trix_period.Value

    @TrixPeriod.setter
    def TrixPeriod(self, value):
        self._trix_period.Value = value

    @property
    def SignalPeriod(self):
        return self._signal_period.Value

    @SignalPeriod.setter
    def SignalPeriod(self, value):
        self._signal_period.Value = value

    @property
    def CandleType(self):
        return self._candle_type.Value

    @CandleType.setter
    def CandleType(self, value):
        self._candle_type.Value = value

    def OnStarted2(self, time):
        super(color_zerolag_trix_osma_strategy, self).OnStarted2(time)

        trix = Trix()
        trix.Length = self.TrixPeriod
        signal = ExponentialMovingAverage()
        signal.Length = self.SignalPeriod

        self.SubscribeCandles(self.CandleType) \
            .Bind(trix, signal, self.ProcessCandle) \
            .Start()

    def ProcessCandle(self, candle, trix_value, signal_value):
        if candle.State != CandleStates.Finished:
            return

        osma = float(trix_value) - float(signal_value)
        self._count += 1

        if self._count < 3:
            self._prev_prev_osma = self._prev_osma
            self._prev_osma = osma
            return

        turn_up = self._prev_osma < self._prev_prev_osma and osma > self._prev_osma
        turn_down = self._prev_osma > self._prev_prev_osma and osma < self._prev_osma

        if turn_up and self.Position <= 0:
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
        elif turn_down and self.Position >= 0:
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()

        self._prev_prev_osma = self._prev_osma
        self._prev_osma = osma

    def OnReseted(self):
        super(color_zerolag_trix_osma_strategy, self).OnReseted()
        self._prev_osma = 0.0
        self._prev_prev_osma = 0.0
        self._count = 0

    def CreateClone(self):
        return color_zerolag_trix_osma_strategy()

File: API/1901-2000/1999_Color_Zerolag_TriX_OSMA/CS/ColorZerolagTrixOsmaStrategy.cs
using System;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Strategy based on a zero-lag TRIX OSMA oscillator.
/// Uses TRIX direction changes for trend reversal signals.
/// </summary>
public class ColorZerolagTrixOsmaStrategy : Strategy
{
	private readonly StrategyParam<int> _trixPeriod;
	private readonly StrategyParam<int> _signalPeriod;
	private readonly StrategyParam<DataType> _candleType;

	private decimal _prevOsma;
	private decimal _prevPrevOsma;
	private int _count;

	public int TrixPeriod { get => _trixPeriod.Value; set => _trixPeriod.Value = value; }
	public int SignalPeriod { get => _signalPeriod.Value; set => _signalPeriod.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public ColorZerolagTrixOsmaStrategy()
	{
		_trixPeriod = Param(nameof(TrixPeriod), 14)
			.SetGreaterThanZero()
			.SetDisplay("TRIX Period", "TRIX calculation period", "Indicator");

		_signalPeriod = Param(nameof(SignalPeriod), 9)
			.SetGreaterThanZero()
			.SetDisplay("Signal Period", "Signal line EMA period", "Indicator");

		_candleType = Param(nameof(CandleType), TimeSpan.FromHours(4).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		=> [(Security, CandleType)];

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_prevOsma = 0;
		_prevPrevOsma = 0;
		_count = 0;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var trix = new Trix { Length = TrixPeriod };
		var signal = new ExponentialMovingAverage { Length = SignalPeriod };

		SubscribeCandles(CandleType)
			.Bind(trix, signal, ProcessCandle)
			.Start();
	}

	private void ProcessCandle(ICandleMessage candle, decimal trixValue, decimal signalValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		var osma = trixValue - signalValue;
		_count++;

		if (_count < 3)
		{
			_prevPrevOsma = _prevOsma;
			_prevOsma = osma;
			return;
		}

		// Buy when OSMA turns up
		var turnUp = _prevOsma < _prevPrevOsma && osma > _prevOsma;
		// Sell when OSMA turns down
		var turnDown = _prevOsma > _prevPrevOsma && osma < _prevOsma;

		if (turnUp && Position <= 0)
		{
			if (Position < 0) BuyMarket();
			BuyMarket();
		}
		else if (turnDown && Position >= 0)
		{
			if (Position > 0) SellMarket();
			SellMarket();
		}

		_prevPrevOsma = _prevOsma;
		_prevOsma = osma;
	}
}
